euphonic/fast_adaptive_broadening.py

```diff
@@ -0,0 +1,162 @@
+        Scalar float. Acceptable error for gaussian approximations, defined
+        as the absolute difference between the areas of the true and
+        approximate gaussians.
+
+    Returns
+    -------
+    spectrum
+        Float ndarray of shape (dos_bins - 1,) containing broadened spectrum
+        ydata
+    """
+    x = np.ravel(x)
+    widths = np.ravel(widths)
+    weights = np.ravel(weights)
+
+    # determine spacing value for mode_width samples given desired error level
+    # coefficients determined from a polynomial fit to plot of
+    # error vs spacing value
+    spacing = np.polyval([ 612.7, -122.7, 15.40, 1.0831], adaptive_error)
+
+    # bins must be regularly spaced
+    bin_width = bins[1]-bins[0]
+
+    n_kernels = int(
+        np.ceil(np.log(max(widths)/min(widths))/np.log(spacing)))
+    width_samples = spacing**np.arange(n_kernels+1)*min(widths)
+    # Determine frequency range for gaussian kernel, the choice of
+    # 3*max(sigma) is arbitrary but tuned for acceptable error/peformance
+    x_range = 3*max(widths)
+    kernel_npts_oneside = np.ceil(x_range/bin_width)
+    kernels = gaussian(np.arange(-kernel_npts_oneside,
+                                 kernel_npts_oneside+1, 1)*bin_width,
+                       width_samples[:, np.newaxis])*bin_width
+    kernels_idx = np.searchsorted(width_samples, widths)
+
+    lower_coeffs = find_coeffs(spacing)
+    spectrum = np.zeros(len(bins)-1)
+
+    # start loop over kernels from 1 as points with insert-position 0
+    # lie outside of bin range but first include points
+    # where mode_widths = min(mode_width_samples)
+    masked_block = (widths == min(width_samples))
+    width_factors = widths[masked_block]/width_samples[0]
+    lower_weights = (np.polyval(lower_coeffs, width_factors))*weights[masked_block]
+    hist, _ = np.histogram(x[masked_block], bins=bins,
+                           weights=lower_weights/bin_width)
+    spectrum += convolve(hist, kernels[0], mode="same", method="fft")
+
+    for i in range(1, len(width_samples)):
+        masked_block = (kernels_idx == i)
+        width_factors = widths[masked_block]/width_samples[i-1]
+        lower_mix = np.polyval(lower_coeffs, width_factors)
+        lower_weights = lower_mix * weights[masked_block]
+
+        if i == 1:
+            hist, _ = np.histogram(x[masked_block], bins=bins,
+                                   weights=lower_weights/bin_width)
+        else:
+            mixing_weights = np.concatenate((upper_weights_prev,
+                                             lower_weights))
+            hist_x = np.concatenate((x_prev, x[masked_block]))
+            hist, _ = np.histogram(hist_x, bins=bins,
+                                   weights=mixing_weights/bin_width)
+
+        x_prev = x[masked_block]
+        upper_weights_prev = weights[masked_block] - lower_weights
+
+        spectrum += convolve(hist, kernels[i-1], mode="same", method="fft")
+
+        if i == len(width_samples)-1:
```

Comment:
I think we can eliminate this logic by instead iterating i up to len(width_samples) + 1). The leftover upper_weights should get concatenated onto zero new data (as the masked_block will be empty).